### ISSUE TITLE
ceph : run rgw with different daemon name for a deployment

### ISSUE THREAD

**Issue by thotz** [rook/rook#6943]
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
Currently when creating a cephobjectstore instance and deployment will be generated, in which a pod with rgw daemon is running on it. If u spin-off another pod out of deployment it rgw daemon running on the second will have the same daemon. Using this feature each RGW instance from the same deployment will have a different name

**What is use case behind this feature:**
Horizontal Pod scaling is the feature in k8s used to scale pods on deployment or Replica set based on custom metrics. In the current scenario even though pods in the ceph object store deployment autoscale, there not much use since all the rgw instance has the same name. There is not much use since s3 clients end up communicating with one instance of rgw, even though other rgw servers are running(no effective scaling of rgw servers).

**Comment by thotz** [comment]
@leseb and @travisn @BlaineEXE: As discussed in the mail chain, I tried to use the custom name by giving `--name` in makeDaemonContainer(), RGW pods are not coming up. Following error is shown in rgw pod log
```
debug 2021-01-15T10:55:29.499+0000 7eff7b0c9700 -1 monclient(hunting): handle_auth_bad_method server allowed_methods [2] but i only support [2,1]
failed to fetch mon config (--no-mon-config to skip)
```

I guess `--name` should match `--id`, it worked with following changes instead of custom name
```diff
diff --git a/pkg/operator/ceph/object/spec.go b/pkg/operator/ceph/object/spec.go
index 71147c487..c8fdee404 100644
--- a/pkg/operator/ceph/object/spec.go
+++ b/pkg/operator/ceph/object/spec.go
@@ -156,6 +156,7 @@ func (c *clusterConfig) makeDaemonContainer(rgwConfig *rgwConfig) v1.Container {
                Args: append(
                        controller.DaemonFlags(c.clusterInfo, c.clusterSpec,
                                strings.TrimPrefix(generateCephXUser(rgwConfig.ResourceName), "client.")),
+                       cephconfig.NewFlag("name", generateCephXUser(rgwConfig.ResourceName)),
                        "--foreground",
                        cephconfig.NewFlag("rgw frontends", fmt.Sprintf("%s %s", rgwFrontendName, c.portString())),
                        cephconfig.NewFlag("host", controller.ContainerEnvVarReference(k8sutil.PodNameEnvVar)),
``` 
In `vstart.sh` it uses `--name` instead of `id` and it matches with `cephxuser`
Am I missing something?

Adding @mattbenjamin @cbodley as well

**Comment by travisn** [comment]
You're saying it worked by passing `name`? If it worked, what is the remaining question? The error makes it look like the cephx key doesn't exist.

I thought the `name` would need to be the pod name in order to give a unique name for each rgw daemon that starts up. The `name` would not match the cephx key, but if the `id` matches the cephx key, that seems like it would work. 

To clarify, maybe you could share the rgw daemon name seen in `ceph status`, and the full args in the pod spec from `kubectl get pod <rgw> -o yaml`

**Comment by leseb** [comment]
The error message is incomplete, the difference between (at least that I'm aware of) `name` and `id` is:

* `--name`: `client.foo`
* `--id`: `foo`

I believe name works because it matches the name inside the keyring file, e,g: `client.rgw.my.store.a`?

**Comment by thotz** [comment]
> You're saying it worked by passing `name`? If it worked, what is the remaining question? The error makes it look like the cephx key doesn't exist.
> 
> I thought the `name` would need to be the pod name in order to give a unique name for each rgw daemon that starts up. The `name` would not match the cephx key, but if the `id` matches the cephx key, that seems like it would work.
> 
> To clarify, maybe you could share the rgw daemon name seen in `ceph status`, and the full args in the pod spec from `kubectl get pod <rgw> -o yaml`

If my understanding is correct we cannot have `custom name` different from `id`, as mentioned in @leseb previous comment. If `--id=foo` then `--name` can only `client.foo` otherwise rgw pod is not coming up

**Comment by leseb** [comment]
> > You're saying it worked by passing `name`? If it worked, what is the remaining question? The error makes it look like the cephx key doesn't exist.
> > I thought the `name` would need to be the pod name in order to give a unique name for each rgw daemon that starts up. The `name` would not match the cephx key, but if the `id` matches the cephx key, that seems like it would work.
> > To clarify, maybe you could share the rgw daemon name seen in `ceph status`, and the full args in the pod spec from `kubectl get pod <rgw> -o yaml`
> 
> If my understanding is correct we cannot have `custom name` different from `id`, as mentioned in @leseb previous comment. If `--id=foo` then `--name` can only `client.foo` otherwise rgw pod is not coming up

Do you have a prototype already?

**Comment by thotz** [comment]
@leseb: So as of now rook creates RGW deployment with a `cephxuser` and rgw daemon is running using that user. Each daemon requires one cephxuser and `name` or `id` represents `cephxuser`. HPA just scales the pods in the Deployment or more like cloning the existing or running pods. 
IMHO (mostly based on assumption, not sure how much is feasible). Rook needs to watch over HPA CRs, then it needs to create to required no of cephxuser based on `max pod` count and `deployment` need to scaling when during its creation. Say if the deployment is `rook-ceph-rgw-my-store-a` and `max pod` count is 5. it needs to create cephxuser `rgw.mystore.a.{1..4}` and when a  pod is created using init container we expose one of the users as an environment variable and pass it as the `id` for daemon container.

Any thoughts @leseb @travisn @BlaineEXE ???

**Comment by travisn** [comment]
@thotz The challenge is that the pod spec must be exactly the same for all 5 instances. In order to generate the cephx keys for all 5 in advanced, we need to use both unique and predictable names. I believe the only unique name for the pod will be the pod name, which is not predictable. Have you seen if the pod scaling has a more predictable pod name or some other property that we could use from the init container?

**Comment by thotz** [comment]
@travisn: Currently, I don't have set up now to cross-check. AFAIR, I didn't see anything special for pods generated by autoscaler, it has a similar spec as the original pod and  similar pod hash value

**Comment by mattbenjamin** [comment]
It seems we need a way to provide a (predictable?) parameter to the
pod?  like a number from 1..n?

Matt

On Mon, Jan 25, 2021 at 2:09 PM Travis Nielsen <notifications@github.com> wrote:
>
> @thotz The challenge is that the pod spec must be exactly the same for all 5 instances. In order to generate the cephx keys for all 5 in advanced, we need to use both unique and predictable names. I believe the only unique name for the pod will be the pod name, which is not predictable. Have you seen if the pod scaling has a more predictable pod name or some other property that we could use from the init container?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.



-- 

Matt Benjamin
Red Hat, Inc.
315 West Huron Street, Suite 140A
Ann Arbor, Michigan 48103

http://www.redhat.com/en/technologies/storage

tel.  734-821-5101
fax.  734-769-8938
cel.  734-216-5309

**Comment by travisn** [comment]
Per separate email thread, the proposal is to use the RADOS client session ID to register the RGW daemon and use it as the unique identifier for each RGW pod. This removes the requirement to pass different arguments to the pods, which won't be possible since they all have the same pod spec.

**Comment by thotz** [comment]
Pasting link for ceph PR for reference https://github.com/ceph/ceph/pull/39380

**Comment by travisn** [comment]
@thotz What is the latest thought on this?

**Comment by thotz** [comment]
We can close this one. The required fix merged upstream and Rook also consumed it in 20c302f72d. Hence closing this one
